src/ScheduledTask/ProcessNotificationsHandler.php

```diff
@@ -0,0 +1,197 @@
+ *
+ * Copyright (c) 2020 Adyen B.V.
+ * This file is open source and available under the MIT license.
+ * See the LICENSE file for more info.
+ *
+ * Author: Adyen <shopware@adyen.com>
+ */
+
+namespace Adyen\Shopware\ScheduledTask;
+
+use Adyen\Shopware\AdyenPaymentShopware6;
+use Adyen\Shopware\Entity\Notification\NotificationEntity;
+use Adyen\Shopware\NotificationProcessor\NotificationProcessorFactory;
+use Adyen\Shopware\Service\NotificationService;
+use Adyen\Shopware\Service\Repository\OrderRepository;
+use Psr\Log\LoggerAwareTrait;
+use Shopware\Core\Checkout\Order\Aggregate\OrderTransaction\OrderTransactionStateHandler;
+use Shopware\Core\Checkout\Payment\PaymentMethodEntity;
+use Shopware\Core\Framework\Context;
+use Shopware\Core\Framework\DataAbstractionLayer\EntityRepositoryInterface;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Criteria;
+use Shopware\Core\Framework\DataAbstractionLayer\Search\Filter\EqualsFilter;
+use Shopware\Core\Framework\MessageQueue\ScheduledTask\ScheduledTaskHandler;
+use Shopware\Core\Framework\Plugin\Util\PluginIdProvider;
+
+class ProcessNotificationsHandler extends ScheduledTaskHandler
+{
+    use LoggerAwareTrait;
+
+    /**
+     * @var NotificationService
+     */
+    private $notificationService;
+    /**
+     * @var OrderRepository
+     */
+    private $orderRepository;
+    /**
+     * @var OrderTransactionStateHandler
+     */
+    private $transactionStateHandler;
+    /**
+     * @var EntityRepositoryInterface
+     */
+    private $paymentMethodRepository;
+    /**
+     * @var PluginIdProvider
+     */
+    private $pluginIdProvider;
+    /**
+     * @var array|null
+     */
+    private $adyenPaymentMethodIds = null;
+
+    private const MAX_ERROR_COUNT = 3;
+
+    public function __construct(
+        EntityRepositoryInterface $scheduledTaskRepository,
+        NotificationService $notificationService,
+        OrderRepository $orderRepository,
+        OrderTransactionStateHandler $transactionStateHandler,
+        EntityRepositoryInterface $paymentMethodRepository,
+        PluginIdProvider $pluginIdProvider
+    ) {
+        parent::__construct($scheduledTaskRepository);
+        $this->notificationService = $notificationService;
+        $this->orderRepository = $orderRepository;
+        $this->transactionStateHandler = $transactionStateHandler;
+        $this->paymentMethodRepository = $paymentMethodRepository;
+        $this->pluginIdProvider = $pluginIdProvider;
+    }
+
+    public static function getHandledMessages(): iterable
+    {
+        return [ ProcessNotifications::class ];
+    }
+
+    public function run(): void
+    {
+        $context = Context::createDefaultContext();
+        $notifications = $this->notificationService->getScheduledUnprocessedNotifications();
+
+        foreach ($notifications->getElements() as $notification) {
+            /** @var NotificationEntity $notification */
+
+            $this->markAsProcessing($notification->getId());
```

Comment:
Not sure if this is going to be an issue or not just thinking about an edge case:

If for any reason the cron job is running multiple times then it might happen that we pick up the same 100 notifications to be processed. Then maybe just to be more fail safe we could check also in this foreach if the notification is in progress or is already done and then we can still skip it.

I can imagine later when we process more notification types the required time of processing 1 notification can go up and can happen that 1 minute won't be enough for the processing.

Please also think about it maybe there are other edge cases.

Reply:
The `getScheduledUnprocessedNotifications` function will only pick up notifications that have `processing=0` AND `done=0` and it is sorted by `scheduled_processing_time`.
When we mark as processing, the notification already does not meet this criteria anymore, therefore it cannot be picked up again.
In the case of an Exception, when we requeue the notification by updating the `scheduled_processing_time`, that way it is picked up last after all the other notifications that are already in the table.

Reply:
okay, let's keep it like this for now and if we see that this is happening that cron job is running multiple times then we can think about it again